File: exercise/week02/c#/Shift/Vertical.cs
namespace Shift;

/*public static class Vertical
{
    public static int WhichFloor(string signalStream)
    {
        List<Tuple<char, int>> val = []; //useless to use the Tuple, you can sum directly the values

        for (int i = 0; i < signalStream.Length; i++)
        {
            var c = signalStream[i];

            if (signalStream.Contains("🧝")) contains the emoji (50% of cases)
            {
                int j;
                if (c == ')') j = 3;
                else j = -2; the emoji is not an instruction, so it should be ignored, not counted as -2

                val.Add(new Tuple<char, int>(c, j));
            }
            else if (!signalStream.Contains("🧝")) doesnt contain the emoji (the other 50%)
            {
                val.Add(new Tuple<char, int>(c, c == '(' ? 1 : -1));
            }
            else val.Add(new Tuple<char, int>(c, c == '(' ? 42 : -2)); never executed
        }

        int result = 0;
        foreach (var kp in val)
        {
            result += kp.Item2;
        }
        return result;
    }
}*/

public static class Vertical
{
    private const char UpSignal = '(';
    private const char DownSignal = ')';
    private const string AnomalySignal = "🧝";

    public static int WhichFloor(string signalStream)
    {
        //Console.WriteLine($"Contains anomaly? {signalStream.Contains(AnomalySignal)}");
        //Console.WriteLine($"Signal length: {signalStream.Length}");
        return signalStream.Contains(AnomalySignal)
            ? WhichFloorUnderAnomaly(signalStream)
            : WhichFloorNormally(signalStream);
    }

    private static int WhichFloorNormally(string signalStream)
    {
        var floor = 0;
        foreach (var signal in signalStream)
        {
            floor += signal switch
            {
                UpSignal => 1,
                DownSignal => -1,
                _ => 0 //in case of any other character, we ignore it and add 0 to the floor
            };
        }
        return floor;
    }

    private static int WhichFloorUnderAnomaly(string signalStream)
    {
        var floor = 0;
        foreach (var signal in signalStream)
        {
            if (signal == DownSignal) floor += 3;
            else if (signal == UpSignal) floor -= 2;
        }
        return floor;
    }
}

File: exercise/week02/php/src/Shift/Vertical.php
<?php

namespace Shift;

/*class Vertical
{
    public static function whichFloor(string $signalStream): int
    {
        $val = [];

        for ($i = 0; $i < strlen($signalStream); $i++) {
            $c = $signalStream[$i];

            if (strpos($signalStream, "🧝")) {
                $j = ($c === ')') ? 3 : -2;
                $val[] = [$c, $j];
            } else if (strpos($signalStream, "🧝") === false) {
                $val[] = [$c, ($c === '(') ? 1 : -1];
            } else {
                $val[] = [$c, ($c === '(') ? 42 : -2];
            }
        }

        $result = 0;
        foreach ($val as $kp) {
            $result += $kp[1];
        }

        return $result;
    }
}*/

class Vertical
{
    private const UP_SIGNAL = '(';
    private const DOWN_SIGNAL = ')';
    private const ANOMALY_SIGNAL = '🧝';

    public static function whichFloor(string $signalStream): int
    {
        return str_contains($signalStream, self::ANOMALY_SIGNAL)
            ? self::whichFloorUnderAnomaly($signalStream)
            : self::whichFloorNormally($signalStream);
    }

    private static function whichFloorNormally(string $signalStream): int
    {
        $floor = 0;
        //mb_str_split is used to handle multibyte characters correctly, such as emojis
        //with str_split, the emoji would be split into multiple characters, which is not what we want
        foreach (mb_str_split($signalStream) as $signal) {
            if ($signal === self::UP_SIGNAL) $floor += 1;
            elseif ($signal === self::DOWN_SIGNAL) $floor -= 1;
            else $floor += 0; 
        }
        return $floor;
    }

    private static function whichFloorUnderAnomaly(string $signalStream): int
    {
        $floor = 0;
        foreach (mb_str_split($signalStream) as $signal) {
            if ($signal === self::DOWN_SIGNAL) $floor += 3;
            elseif ($signal === self::UP_SIGNAL) $floor -= 2;
        }
        return $floor;
    }
}